readme.md
#### rspec_flake
[![Gem Version](https://badge.fury.io/rb/rspec_flake.svg)](http://badge.fury.io/rb/rspec_flake)
[![Dependency Status](https://gemnasium.com/bootstraponline/rspec_flake.svg?nocache)](https://gemnasium.com/bootstraponline/rspec_flake)
[![Build Status](https://travis-ci.org/bootstraponline/rspec_flake.svg?branch=master)](https://travis-ci.org/bootstraponline/rspec_flake/builds)
[![Coverage Status](https://coveralls.io/repos/bootstraponline/rspec_flake/badge.svg?nocache)](https://coveralls.io/r/bootstraponline/rspec_flake)

Measure RSpec test flakiness and performance

#### Design Notes

`rspec` doesn't enforce unique test names, also test suite names may be reused
in different files. The only reliable way to identify a test or suite is by
location.

A nokogiri SAX parser is used to read the junit xml into a hash keyed on
location. After that the hash is rewritten and passed to XmlSimple for
serilization to xml.

The value of `errors` is hardcoded to 0 in `jarjuf` because rspec records
failures. Errors are considered a type of failure.

#### Example output

```
$ rspec_flake 2 rspec
Running: "rspec" --format JUnit --out "/rspec_simple_example/tmp/0.xml"
Running: "rspec" --format JUnit --out "/rspec_simple_example/tmp/1.xml"
a a 1 - runs: 2 - failures: 0 - avg time: 0.0 - ./spec/a_spec.rb:6
a a 2 - runs: 2 - failures: 0 - avg time: 0.0 - ./spec/a_spec.rb:10
a a 3 - runs: 2 - failures: 0 - avg time: 0.0 - ./spec/a_spec.rb:13
b b 3 - runs: 2 - failures: 0 - avg time: 0.0 - ./spec/b_spec.rb:6
b b 4 - runs: 2 - failures: 0 - avg time: 0.0 - ./spec/b_spec.rb:9
```

#### Parallel

When running [tests in parallel](https://github.com/grosser/parallel_tests),
the output can be [saved by env number](http://elementalselenium.com/tips/57-junit-xml)
in the `.rspec_parallel` options file

`--out junit/xml_<%= ENV['TEST_ENV_NUMBER'] %>.xml`

There's also the `.rspec` options file and flags such as `--format documentation`
can be passed directly to rspec_flake.

#### Google Sheets Note

The file time output is tab delimited and can be posted directly into
a Google Sheets document using Command/Control + Shift + V.

#### Run tests

- `thor spec`


release_notes.md
#### v0.2.3 2015-06-19

- [4e54fd0](https://github.com/appium/rspec_flake/commit/4e54fd0815191d2dc0b1ea64a11b031907297eff) Release 0.2.3
- [926424a](https://github.com/appium/rspec_flake/commit/926424adc1ab006893970d4663ac68c0b516419e) Fix Thorfile
- [d17c0ec](https://github.com/appium/rspec_flake/commit/d17c0ec7bc3a3d5a9f12609a6216c4908e6af94f) Test RSpecFlake::Runner.run_tests
- [7f07dde](https://github.com/appium/rspec_flake/commit/7f07dde4d10741920248ce0c34fc08485544e8cc) Cache bundler
- [e738900](https://github.com/appium/rspec_flake/commit/e738900656ec329a0bfe52e5bc18683372e808af) Setup coveralls and gemnasium
- [f355696](https://github.com/appium/rspec_flake/commit/f3556966b1274147ec7ddf48b6c88ed465b27fe5) Update readme.md
- [3d01968](https://github.com/appium/rspec_flake/commit/3d01968c16e275a4d258c3acb8246f056438ba3c) Create .rspec_parallel
- [5224921](https://github.com/appium/rspec_flake/commit/522492199894e029bb2a8d68b774bea5844326f3) Support parallel_rspec
- [3ab091e](https://github.com/appium/rspec_flake/commit/3ab091ee8ca45c31614693a7035346db0f01d688) Update specs
- [ff493b1](https://github.com/appium/rspec_flake/commit/ff493b15d231f9437f93225bd641bee21368b235) Fix specs
- [7c1d528](https://github.com/appium/rspec_flake/commit/7c1d528b0d1bbacd9e70c4828ac051f6328131ec) Clarify docs
- [b596cd0](https://github.com/appium/rspec_flake/commit/b596cd0a342a614a9d4c130050146494be138d2f) Fix specs and update docs
- [3990e6e](https://github.com/appium/rspec_flake/commit/3990e6e52c1ade662fa27f7ad6440d711ea79466) Fix specs
- [9416a70](https://github.com/appium/rspec_flake/commit/9416a7036e77b04748f465fcd94a2040d3f9fe1d) Update file time output
- [8ffa7ab](https://github.com/appium/rspec_flake/commit/8ffa7ab7e0fbf2fefda2f9b95c1bc758db5346db) Add file time calculation
- [78e2623](https://github.com/appium/rspec_flake/commit/78e2623baf6f4033664c583b932cf184d9f7eb57) Add general merge method


#### v0.2.2 2015-01-19

- [91c768f](https://github.com/appium/rspec_flake/commit/91c768f69af0bc4d7fe7bbd085c00c6e80dd5c7d) Release 0.2.2
- [e58da69](https://github.com/appium/rspec_flake/commit/e58da69ba710070e621b98e5f7b7b28b53c9ac5c) Fix arg parsing


#### v0.2.1 2015-01-19

- [f8abf03](https://github.com/appium/rspec_flake/commit/f8abf030b53818493816fd901f95aef8a727a664) Release 0.2.1
- [add98a0](https://github.com/appium/rspec_flake/commit/add98a0fad8c60c8938f050ac3d65dc6109a0a5c) Don't quote command
- [7b3ff90](https://github.com/appium/rspec_flake/commit/7b3ff907c79cc5f8de34787da1b89e26f975ab71) Update readme.md


#### v0.2.0 2015-01-19

- [abfbc14](https://github.com/appium/rspec_flake/commit/abfbc142d50404d608e49a3f009363f802874aee) Release 0.2.0
- [c7c14b6](https://github.com/appium/rspec_flake/commit/c7c14b668e3687b277c8706eb3e7b8e4ed627230) Save merged.xml outside of tmp
- [c647298](https://github.com/appium/rspec_flake/commit/c64729842cbadde8efac46d621c682aec1663fec) Update readme.md
- [b61893b](https://github.com/appium/rspec_flake/commit/b61893b73ff568ef251196621fecaf8451306154) Release 0.0.2
- [ce71138](https://github.com/appium/rspec_flake/commit/ce7113804f64b1a85c4a06171c8021e572e0a02c) Fix scope problems
- [bd53649](https://github.com/appium/rspec_flake/commit/bd536499eb790c5193fa81d60814679e95b9d4e0) Update stats
- [c4219a9](https://github.com/appium/rspec_flake/commit/c4219a9caed77003e9a299588f1f6c2146945134) Format code
- [7fa073d](https://github.com/appium/rspec_flake/commit/7fa073d7038f94c0127ac754238df5011d22d70d) Write merged.xml by default
- [0baecf7](https://github.com/appium/rspec_flake/commit/0baecf783ae99b1e2ad671d7e05936f4c05a510c) Refactor merge test
- [d99cd45](https://github.com/appium/rspec_flake/commit/d99cd458f01efe96074c1751cccd30e52d2e70e9) Store failure cdata
- [51acdf6](https://github.com/appium/rspec_flake/commit/51acdf6393a0d52d17df064e9d10fe805c5439ac) Refactor test object
- [c20efdd](https://github.com/appium/rspec_flake/commit/c20efdd997eeb47382b4fb4ef48ed027e1573d47) Add merge feature
- [11bcc70](https://github.com/appium/rspec_flake/commit/11bcc70811373ca34aa350ab3501118f348f50c1) Fix CDATA seralization
- [46646e8](https://github.com/appium/rspec_flake/commit/46646e80eda3179cb073b16b9d566f8f6fb24f74) Save failure cdata as content
- [9b927ad](https://github.com/appium/rspec_flake/commit/9b927ad3503d262e85895d66e5781190ae582f16) Add failure support
- [663591d](https://github.com/appium/rspec_flake/commit/663591d6ffd4f3ea46118d72623016f4e42e4b4b) Add design notes
- [dc715c2](https://github.com/appium/rspec_flake/commit/dc715c224e0ca7d206f13249d3827359091268f6) Update for Travis CI
- [73ceeec](https://github.com/appium/rspec_flake/commit/73ceeec21ae764885f913464b5b313c08ecae5c8) Use XmlSimple and add specs
- [f268a1e](https://github.com/appium/rspec_flake/commit/f268a1e77c35d8e391eb668c4cb0ccf9805ee618) Init rspec_flake
- [bc3375e](https://github.com/appium/rspec_flake/commit/bc3375e4a8f68d632e09df981590eeeb9532c43f) Update adb test
- [0c18668](https://github.com/appium/rspec_flake/commit/0c18668ae7c3f394fe4b8544a2b1cbca22904407) Add adb test
- [171333c](https://github.com/appium/rspec_flake/commit/171333c172d26a3ef2da0165170894a66170cff2) Capture debug logs from appium server
- [7177da4](https://github.com/appium/rspec_flake/commit/7177da4e14f711710bfec1d53e260dabcd92e3d0) Add flaky.txt support to run/one_test.rb


#### v0.1.1 2014-05-19

- [98f3c87](https://github.com/appium/rspec_flake/commit/98f3c87c1caa7c64b1d1da26f43642f4bdef04d1) Release 0.1.1
- [62d4278](https://github.com/appium/rspec_flake/commit/62d427893f19fea1729a5c3c6f79ce72e1a11919) Fix test name
- [963978f](https://github.com/appium/rspec_flake/commit/963978f041f2198dda07779505a058e971e13273) Default to no video
- [dd0ea3a](https://github.com/appium/rspec_flake/commit/dd0ea3a99180e3c687ce67aae6cbf6abe2dac163) Fix paren


#### v0.1.0 2014-04-30

- [f8a79a4](https://github.com/appium/rspec_flake/commit/f8a79a47fe01bd0e1f9e668b0a97d7d9ba6d41b7) Release 0.1.0

spec/data/merge.rb
=begin
<testsuite errors="0" failures="0" skipped="0" tests="4" time="0.001283" timestamp="2015-01-15T10:25:40-05:00">
  <testsuite location="./spec/a_spec.rb:1" name="a" tests="2" errors="0" failures="0" skipped="0">
    <testcase name="a a 1" time="0.000177" location="./spec/a_spec.rb:6">

# sax parser
# http://nokogiri.org/Nokogiri/XML/SAX.html
=end

require 'rubygems'
require 'nokogiri'
require 'pry'
require 'awesome_print'

Pry.config.command_prefix = "%" if defined?(Pry)

class ReportParser < Nokogiri::XML::SAX::Document
  def initialize
    reset
  end

  def reset
    @data = {}
    @suite_location = nil
  end

  def result
    @data
  end

  def start_element name, attrs = []
    name = name.to_sym
    testsuite = :testsuite
    root = :testsuites
    location = attrs.to_h['location']

    case name
      when root
        @data[root] ||= {attrs: attrs, testsuite: {}}
      when :testsuite
        @suite_location = location
        @data[root][testsuite][@suite_location] ||= {attrs: attrs, testcase: {}}
      when :testcase
        raise 'testcase not part of a suite' unless @suite_location
        @data[root][testsuite][@suite_location][:testcase][location] ||= attrs
    end
  end

  def end_element name
    @suite_location = nil if name == 'testsuite'
  end
end

def read_as_utf8 path
  File.read(path).encode!('UTF-8', invalid: :replace, undef: :replace, replace: '')
end

xml = read_as_utf8 '0.xml'

parser = Nokogiri::XML::SAX::Parser.new ReportParser.new
parser.document.reset
parser.parse xml

ap parser.document.result


=begin
{
    :testsuites => {
            :attrs => [
            [0] [
                [0] "errors",
                [1] "0"
            ],
            [1] [
                [0] "failures",
                [1] "0"
            ],
            [2] [
                [0] "skipped",
                [1] "0"
            ],
            [3] [
                [0] "tests",
                [1] "4"
            ],
            [4] [
                [0] "time",
                [1] "0.001283"
            ],
            [5] [
                [0] "timestamp",
                [1] "2015-01-15T10:25:40-05:00"
            ]
        ],
        :testsuite => {
            "./spec/a_spec.rb:1" => {
                   :attrs => [
                    [0] [
                        [0] "location",
                        [1] "./spec/a_spec.rb:1"
                    ],
                    [1] [
                        [0] "name",
                        [1] "a"
                    ],
                    [2] [
                        [0] "tests",
                        [1] "2"
                    ],
                    [3] [
                        [0] "errors",
                        [1] "0"
                    ],
                    [4] [
                        [0] "failures",
                        [1] "0"
                    ],
                    [5] [
                        [0] "skipped",
                        [1] "0"
                    ]
                ],
                :testcase => {
                     "./spec/a_spec.rb:6" => [
                        [0] [
                            [0] "name",
                            [1] "a a 1"
                        ],
                        [1] [
                            [0] "time",
                            [1] "0.000177"
                        ],
                        [2] [
                            [0] "location",
                            [1] "./spec/a_spec.rb:6"
                        ]
                    ],
                    "./spec/a_spec.rb:10" => [
                        [0] [
                            [0] "name",
                            [1] "a a 2"
                        ],
                        [1] [
                            [0] "time",
                            [1] "0.000198"
                        ],
                        [2] [
                            [0] "location",
                            [1] "./spec/a_spec.rb:10"
                        ]
                    ]
                }
            },
            "./spec/b_spec.rb:1" => {
                   :attrs => [
                    [0] [
                        [0] "location",
                        [1] "./spec/b_spec.rb:1"
                    ],
                    [1] [
                        [0] "name",
                        [1] "b"
                    ],
                    [2] [
                        [0] "tests",
                        [1] "2"
                    ],
                    [3] [
                        [0] "errors",
                        [1] "0"
                    ],
                    [4] [
                        [0] "failures",
                        [1] "0"
                    ],
                    [5] [
                        [0] "skipped",
                        [1] "0"
                    ]
                ],
                :testcase => {
                     "./spec/b_spec.rb:6" => [
                        [0] [
                            [0] "name",
                            [1] "b b 3"
                        ],
                        [1] [
                            [0] "time",
                            [1] "0.000112"
                        ],
                        [2] [
                            [0] "location",
                            [1] "./spec/b_spec.rb:6"
                        ]
                    ],
                    "./spec/b_spec.rb:10" => [
                        [0] [
                            [0] "name",
                            [1] "b b 4"
                        ],
                        [1] [
                            [0] "time",
                            [1] "0.000204"
                        ],
                        [2] [
                            [0] "location",
                            [1] "./spec/b_spec.rb:10"
                        ]
                    ]
                }
            }
        }
    }
}
=end

spec/merge_spec.rb
describe 'merge rspec_junit xml' do
  it 'merges successfully' do
    xml_files = %w[0.xml 1.xml 2.xml failure.xml failure2.xml]
    actual    = RSpecFlake.merge_xml input: xml_files.map(&method(:data))
    expected  = expected_merge_xml

    expect(actual).to eq expected
  end

  it 'merges individual xml files' do
    xml_files = %w[3.xml 4.xml]
    actual    = RSpecFlake.merge_individual_xml files: xml_files.map(&method(:data))
    expected  = expected_merge_individual_xml

    expect(actual).to eq(expected)
  end
end


spec/parse_spec.rb
describe 'parse rspec_junit xml' do
  it 'parses success results' do
    actual   = RSpecFlake.parse_xml data '0.xml'
    expected = expected_parsed_0_xml

    expect(actual).to eq expected
  end

  it 'parses failure results' do
    actual   = RSpecFlake.parse_xml data 'failure.xml'
    expected = expected_parsed_failure_xml

    expect(actual).to eq expected
  end
end


lib/rspec_flake/merge.rb
module RSpecFlake
  class << self

    # Merges individual xml reports (from parallel_rspec)
    # into a single combined xml file
    #
    # @example:
    #    merge_individual_xml files: Dir.glob('tmp/*.xml')
    def merge_individual_xml opts={}
      files = opts[:files]
      files = [files] unless files.is_a?(Array)

      files_hash = {}
      files.each { |file| files_hash.merge!(RSpecFlake.parse_xml(file)[:testsuites][:testsuite]) }

      RSpecFlake.hash_to_xml({ testsuites: { testsuite: files_hash } })
    end

    # merge input xml files into a hash
    #
    # input - single path or array of paths to input xml files
    # @return merged hash
    def merge_xml opts={}
      input = opts[:input]
      raise 'input path(s) not provided' unless input
      input = [input] unless input.kind_of?(Array)

      merged = { testsuite: { testcase: {} } }

      # annotate each testcase node with two attributes
      #
      # failures - amount of times this test failed
      # runs     - amount of times this test was executed
      #
      # name/time - saved from the first run of the testcase
      # location - stored for each run

      input.each do |xml_path|
        parsed = parse_xml(xml_path)

        parsed[:testsuites][:testsuite].each do |suite_location, suite_obj|
          suite_obj[:testcase].each do |testcase_location, testcase_obj|
            attrs = testcase_obj[:attrs]
            test  = merged[:testsuite][:testcase][testcase_location] ||= { failures: 0, runs: 0, name: attrs['name'], location: attrs['location'], time: [] }

            if testcase_obj[:failure]
              test[:failures] += 1
              test[:failure]  ||= []
              test[:failure] << { 'content' => cdata(testcase_obj[:failure][:content]) }
            end
            test[:runs] += 1
            test[:time] << attrs['time']
          end
        end
      end

      # transform merge hash keyed on location into simple testcase array
      # for xml conversion
      converted = { testsuite: { testcase: [] } }

      merged[:testsuite][:testcase].each do |testcase_location, testcase_obj|
        converted[:testsuite][:testcase] << testcase_obj
      end

      xml_out converted
    end
  end
end


lib/rspec_flake/convert.rb
module RSpecFlake
  class << self
    def hash_to_xml source_hash
      # top level attribues from root node
      converted             = source_hash[:testsuites][:attrs] || {}
      converted[:testsuite] = []

      source_hash[:testsuites][:testsuite].each do |suite_location, suite_obj|
        suite_with_tests            = suite_obj[:attrs] || {}
        suite_with_tests[:testcase] = []
        suite_obj[:testcase].each do |testcase_location, testcase_obj|
          testcase = testcase_obj[:attrs]
          if testcase_obj[:failure]
            fail_content = testcase_obj[:failure][:content]
            fail_content = cdata fail_content
            # content key must be a string for xml simple
            testcase.merge!(failure: testcase_obj[:failure][:attrs].merge('content' => fail_content))
          end
          suite_with_tests[:testcase] << testcase
        end

        converted[:testsuite] << suite_with_tests
      end

      # ap converted, index: false, indent: 2

      xml_out converted
    end

    def cdata content
      "\n<![CDATA[#{content}]]>\n"
    end

    def xml_out hash
      xml_header = %Q(<?xml version="1.0" encoding="UTF-8"?>\n)
      xml_body   = XmlSimple.xml_out(hash, RootName: 'testsuites')
      # xmlsimple will escape the cdata by default
      xml_body   = EscapeUtils.unescape_html xml_body
      xml_header + xml_body
    end
  end
end


lib/rspec_flake/file_time.rb
module RSpecFlake
  class << self

    # Read in JUnit xml files and output an xml report
    # that lists the time spent per file
    #
    # @example puts RSpecFlake.file_time files: Dir.glob("*.xml")
    def file_time opts={}
      files = opts[:files]
      files = [files] unless files.is_a?(Array)

      file_times = Hash.new 0

      xml = Nokogiri::XML(merge_individual_xml files: files)
      xml.xpath('//testcase').each do |test|
        file_name = test.attr('location').split(':').first
        file_times[file_name] += test.attr('time').to_i
      end

      file_times = file_times.sort_by { |file, time| time }.reverse

      # Paste into Google Sheets with Ctrl/Command + Shift + V
      output = ''
      file_times.each do |file, time|
        time = ChronicDuration.output(time) || '0'
        output += "#{time.ljust(12)}\t#{file}\n"
      end

      output
    end
  end
end


lib/rspec_flake/rspec_flake.rb
# encoding: utf-8
module RSpecFlake
  class Runner
    def self.run_tests opts={}
      io = opts.fetch(:io, $stdout)

      count = opts[:count]
      raise 'count is required and must be a number' unless count && count.to_s.match(/\d+/)
      count = count.to_i

      command = opts[:command]
      raise "command is required and must start with rspec. #{command}" unless command && command.include?('rspec')

      tmp_path = File.expand_path File.join Dir.pwd, 'tmp'
      FileUtils.rm_rf tmp_path
      FileUtils.mkdir_p tmp_path

      xml_files = []

      count.times do |iteration|
        out_file = File.expand_path File.join(tmp_path, iteration.to_s + '.xml')
        xml_files << out_file
        spawn_command = %Q(#{command} --format JUnit --out "#{out_file}")
        io.puts "Running: #{spawn_command}"
        Process::waitpid(POSIX::Spawn::spawn(spawn_command))
      end

      merge_path = File.join(tmp_path, '..', 'merged.xml')
      File.open(merge_path, 'w') do |file|
        xml = RSpecFlake.merge_xml input: xml_files
        file.write xml
        io.puts RSpecFlake.stats_from_merge_xml xml
      end
    end
  end
end


lib/rspec_flake/parse.rb
=begin
<testsuite errors="0" failures="0" skipped="0" tests="4" time="0.001283" timestamp="2015-01-15T10:25:40-05:00">
  <testsuite location="./spec/a_spec.rb:1" name="a" tests="2" errors="0" failures="0" skipped="0">
    <testcase name="a a 1" time="0.000177" location="./spec/a_spec.rb:6">

# sax parser
# http://nokogiri.org/Nokogiri/XML/SAX.html
=end

module RSpecFlake
  class << self
    class ReportParser < Nokogiri::XML::SAX::Document
      def initialize
        reset
      end

      def reset
        @data              = {}
        @suite_location    = nil
        @testcase_location = nil
        @failure           = nil
      end

      def result
        @data
      end

      def start_element name, attrs = []
        name      = name.to_sym
        attrs     = attrs.to_h
        testsuite = :testsuite
        root      = :testsuites
        location  = attrs['location']

        case name
          when root
            @data[root] ||= { attrs: attrs, testsuite: {} }
          when :testsuite
            @suite_location                         = location
            @data[root][testsuite][@suite_location] ||= { attrs: attrs, testcase: {} }
          when :testcase
            raise 'testcase not part of a suite' unless @suite_location
            @testcase_location                                           = location
            @data[root][testsuite][@suite_location][:testcase][location] ||= { attrs: attrs }
          when :failure
            raise 'failure not part of a testcase' unless @testcase_location
            @data[root][testsuite][@suite_location][:testcase][@testcase_location].merge!({ failure: { attrs: attrs } })
            @failure = @data[root][testsuite][@suite_location][:testcase][@testcase_location][:failure]
        end
      end

      def cdata_block string
        raise 'cdata not associated with failure' unless @failure
        @failure[:content] = string
        @failure           = nil
      end

      def end_element name
        case name
          when 'testsuite'
            @suite_location = nil
          when 'testcase'
            @testcase_location = nil
        end
      end
    end

    def read_as_utf8 path
      File.read(path).encode!('UTF-8', invalid: :replace, undef: :replace, replace: '')
    end

    # Parse JUnit xml into a Hash keyed on location
    def parse_xml xml_path
      xml = read_as_utf8 xml_path

      parser = @parser ||= Nokogiri::XML::SAX::Parser.new(ReportParser.new)
      parser.document.reset
      parser.parse xml

      parser.document.result
    end
  end
end
